src/rewrite/steward.ts

```diff
@@ -99,29 +110,67 @@ export function handleLogBuyVitalikLegacy(event: LogBuyLegacy): void {
         )
       : BigInt.fromI32(0);
 
+  let newPatronTotalContributed =
+    patronOld.id != "NO_OWNER"
+      ? patron.totalContributed.plus(
+          patron.patronTokenCostScaledNumerator
+            .times(timeSinceLastUpdatePatron)
+            .div(GLOBAL_PATRONAGE_DENOMINATOR)
+            .div(NUM_SECONDS_IN_YEAR_BIG_INT)
+        )
+      : BigInt.fromI32(0);
+  let newPatronTokenCostScaledNumerator = VITALIK_PATRONAGE_NUMERATOR.times(
+    tokenPrice
+  );
+  let oldPatronTotalContributed =
+    patronOld.id != "NO_OWNER"
+      ? patronOld.totalContributed.plus(
+          patronOld.patronTokenCostScaledNumerator
+            .times(timeSinceLastUpdatePatron)
+            .div(GLOBAL_PATRONAGE_DENOMINATOR)
+            .div(NUM_SECONDS_IN_YEAR_BIG_INT)
+        )
+      : BigInt.fromI32(0);
+  let oldPatronTokenCostScaledNumerator = BigInt.fromI32(0);
+
   let newPatronTokenArray = patron.tokens.concat([wildcard.id]);
   let itemIndex = patronOld.tokens.indexOf(wildcard.id);
   let oldPatronTokenArray = patronOld.tokens
     .slice(0, itemIndex)
     .concat(patronOld.tokens.slice(itemIndex + 1, patronOld.tokens.length));
 
-  // Phave 3: set+save values.
+  // Phase 3: set+save values.
 
   patron.lastUpdated = txTimestamp;
   patron.totalTimeHeld = newPatronTotalTimeHeld;
   patron.tokens = newPatronTokenArray;
+  patron.patronTokenCostScaledNumerator = newPatronTokenCostScaledNumerator;
+  patron.totalContributed = newPatronTotalContributed;
   patron.save();
 
   patronOld.lastUpdated = txTimestamp;
   patronOld.totalTimeHeld = oldPatronTotalTimeHeld;
   patronOld.tokens = oldPatronTokenArray;
+  patronOld.patronTokenCostScaledNumerator = oldPatronTokenCostScaledNumerator;
+  patronOld.totalContributed = oldPatronTotalContributed;
   patronOld.save();
 
   wildcard.owner = ownerString;
   wildcard.save();
 }
 
 export function handleLogBuy(event: LogBuy): void {
+  // This is a hack, but after spending 4 hours trying to work out why the
+  //    "0x3ea1ecb8775a37fb797bb79f6c419176f15e35d4" wildcards address is
+  //    still being created (when I bought back vitalik as a test for Simon).
+  //    Theoritically this should be caught by the `isVitalik` section, but for some reason it isn't.
+  if (
+    event.transaction.hash.toHexString() ==
+    "0x8961f14a43c82842d5541da454b44aad64cbbe11b2f9094d6dc191bce3eac2f3"
+  ) {
+    return;
```

Comment:
This is a safety hatch, because this transaction was one of the transactions that I helped Simon fix the issue with Vitalik.

I don't know why (hence the long comment above this) this was actually needed, but without it we record that 0x6d47cf86f6a490c6410fc082fd1ad29cf61492d0 which is the wildcards admin address bought a token and never sold it. see: https://etherscan.io/tx/0x8961f14a43c82842d5541da454b44aad64cbbe11b2f9094d6dc191bce3eac2f3

The perplexing thing is that I thought after removing the `<entity>.save()` when creating new entities that this problem would go away. Since this transaction is caught by the `isVintageVitalik(tokenIdBigInt, event.block.number)` function. [I'll comment it so you can follow the story]

Reply:
Okay great, going to go through this in a bit more detail when I have time tomorrow. Think I will also give you a quick call to double check my understanding :) Going to work on ETHcc stuff quick

Reply:
Sure :) It isn't crazy complicated, but going through it will help!